### DISCUSSION TITLE
Uno Platform Google Signin OAuth2 Setup

### DISCUSSION THREAD

**Discussion by RedEye-Developers** [unoplatform/uno#18868] [Show and tell]
Follow The Steps in This Video To Config Google Api Library.

https://github.com/user-attachments/assets/85bbd5c3-6cad-4a5f-b27f-395d93042622

VIdeo Time : 02:05 = ![Screenshot (21)](https://github.com/user-attachments/assets/804ebf88-8112-4eac-b20a-9814a7849fbe)

Video Time : 02:10 = Must Add Test User Gmail ID to Test The Google Signin in Development other wise in Application Throw Error!

Video Time : 03:30 =  ![Screenshot (17)](https://github.com/user-attachments/assets/1729f96b-2024-400c-b897-b9b66905716e)

Video Time : 03:35 = They have 2 Keystore one for Debug Build another one for Relece Build

If you are going to Run Google SignIn Application in Debug Build Use This Cmd in Command Prompt = keytool -genkeypair -v -keystore debug.keystore -keyalg RSA -keysize 2048 -validity 10000 -alias androiddebugkey

If you are going to Run Google SignIn Application in Relece Build Use This Cmd in Command Prompt = keytool -genkeypair -v -keystore release-key.jks -keyalg RSA -keysize 2048 -validity 10000 -alias release-key

Sample Video to Create Keystore = https://www.youtube.com/watch?v=NQTj5LpjVs4

To View Your KeyStore SHA1 Fingerprint use this Below Cmd in Command Prompt

Sample : keytool -list -v -keystore <keystore_filename> -alias <key_alias>
Debug Keystore : keytool -list -v -keystore ~/.android/debug.keystore -alias androiddebugkey
Relece Keystore : keytool -list -v -keystore release-key.jks -alias release-key

Video Time : 05:45 = ![Screenshot (18)](https://github.com/user-attachments/assets/5207e627-5b12-4191-960f-30c9dd7bf816)

after done in your IDE = ![Screenshot (19)](https://github.com/user-attachments/assets/0fdff44c-c605-4be1-9a2c-ba9c08e983dc)

after done install this package = ![Screenshot (14)](https://github.com/user-attachments/assets/5170e321-83c8-4f45-a64d-d2f2e91e5262)

after install package past this code inside your = MainActivity.Android.cs

```cs
using Android.App;
using Android.Content;
using Android.Gms.Auth.Api;
using Android.Gms.Auth.Api.SignIn;
using Android.Gms.Common.Apis;
using Android.OS;
using Android.Views;
using FitBites.Interface;
using FitBites.Services;

namespace FitBites.Droid;

[Activity(
    MainLauncher = true,
    ConfigurationChanges = global::Uno.UI.ActivityHelper.AllConfigChanges,
    WindowSoftInputMode = SoftInput.AdjustNothing | SoftInput.StateHidden
)]
public class MainActivity : Microsoft.UI.Xaml.ApplicationActivity
{ 
    private readonly int _signInRequestCode = 9001;
    private GoogleSignInOptions _googleSignInOptions;
    private Intent _intent;
    
    private IGoogleLoginService _googleLogin;
    
    protected override void OnCreate(Bundle bundle)
    {
        base.OnCreate(bundle);
        InitializeGoogleSignIn();
    } 
    
    private void InitializeGoogleSignIn()
    {
        GoogleLogin();
        _googleSignInOptions = new GoogleSignInOptions.Builder(GoogleSignInOptions.DefaultSignIn)
            .RequestIdToken(GetString(Resource.String.Web_ClientID))
            .RequestEmail()
            .RequestProfile()
            .RequestScopes(new Scope("https://www.googleapis.com/auth/userinfo.email"), 
                new Scope("https://www.googleapis.com/auth/userinfo.profile"))
            .Build();
        
            GoogleSignInClient googleSignInClient = GoogleSignIn.GetClient(this, _googleSignInOptions);
            _intent = googleSignInClient.SignInIntent;
    }
    
    protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
    {
        base.OnActivityResult(requestCode, resultCode, data);

        if (requestCode == _signInRequestCode)
        {
            var result = Auth.GoogleSignInApi.GetSignInResultFromIntent(data);
            if (result.IsSuccess)
            {
                var account = result.SignInAccount;
                _googleLogin.OnLoginSuccess(account.DisplayName, account.Email, account.PhotoUrl.ToString());
                return;
            }
            _googleLogin.OnLoginFailed(result.Status.StatusMessage);
        }
    }

    public void GoogleLogin()
    {
        _googleLogin = GoogleLogInServics.Instance;
        _googleLogin.OnLogIn += () =>
        {
            StartActivityForResult(_intent, _signInRequestCode);
        };
    }
}


// public static void ShowLogPopup(Activity activity, string logMessage)
// {
//     // Creating an AlertDialog to show logs
//     var builder = new AlertDialog.Builder(activity);
//     builder.SetTitle("Log Message");
//     builder.SetMessage(logMessage);  // Set the log message to display
//     builder.SetPositiveButton("OK", (sender, args) => { });
//
//     // Show the dialog
//     builder.Show();
// }
```

Uno Platform Google SignIn OAuth2 Login Sample

https://github.com/user-attachments/assets/056ddfda-e6d2-404d-9228-a31102a1f1cf

**Comment by RedEye-Developers** [comment]
Forget to say No Need any Backend Server Connections!

**Reply by MartinZikmund** [reply]
This is a very nice guide, would you be willing to turn it into a docs page as a pull request?

**Reply by RedEye-Developers** [reply]
yea sure!

**Reply by morning4coffe-dev** [reply]
Nice! @RSReswin If you need some help getting the PR up, just let me know. I recently did a How to Sign in with Apple button, so you can take an inspiration from that https://github.com/unoplatform/uno/pull/18410

**Reply by DevTKSS** [reply]
@RSReswin as you named this "oAuth2" Setup and wrote the doc page, are you somehow able to help me figuring out, where to put TokenEndpoint + UserInfoEndpoint which are commonly used in oAuth2? Normally social logins are "just" oidc which is different to oAuth2 which I do need now. I am unable to get that working on any device. tryed with android, wasm, windows (not working because my required api requires https://) and desktop. Just so I can make sure, I did not make a simple issue?

**Reply by RedEye-Developers** [reply]
>where to put TokenEndpoint.

i put the Token in String.xml file that comes under the (platform => android => resource => values => string.xml),
after done i just rebuild the project and point the token in code like this : ```RequestIdToken(GetString(Resource.String.Web_ClientID))```

>UserInfoEndpoint

i not use anything like this in Google OAuth2, in google i use Scope to get the user datas like username, profileUrl, email like stuffs 
```
.RequestScopes(new Scope("https://www.googleapis.com/auth/userinfo.email"), 
                new Scope("https://www.googleapis.com/auth/userinfo.profile"))
```

> (not working because my required api requires https://) and desktop

i not use any api stuffs for OAuth2, all code are client side and when user press the google login button and login ```OnActivityResult``` method was fetch the user data and i will send that to api to login the application.

and this OAuth2 for android, it will not work in other platforms.
